torch/distributed/distributed_c10d.py

```diff
@@ -1327,6 +1327,21 @@ def _get_split_source(pg):
 
     return split_from
 
+def _shutdown_backend(pg):
+    """
+    Try to shut down the backend of a process group.
+    Currently, only ProcessGroupNCCL backend is supported.
+    No op for other backends.
+    """
+    backend = None
+    try:
+        backend = pg._get_backend(torch.device("cuda"))
+    except RuntimeError:
+        pass
+    if is_nccl_available() and isinstance(backend, ProcessGroupNCCL):
+        # explictly call shutdown to ensure that NCCL resources are released
+        backend._shutdown()
```

Comment:
This is a new call for destroy_process_group, so it has the potential to introduce hangs in existing code. For instance, whereas previously we would clean up by calling ncclCommAbort in the destructor (which nominally shouldn't hang though it sometimes does), now we will explicitly be waiting for the shutdown call to complete, which _will_ hang if other ranks do not also call it. I think we need to make sure the completion of _shutdown is subject to the standard nccltimeout, and a timeout exception is raised if a rank doesn't also shutdown.

Reply:
Yes, ideally ncclCommAbort should return either success or timeout, but instead it can hang forever. We should be able to add another layer of abstraction to implement this "either success or timeout"

Reply:
To clarify in case my understanding doesn't match what @zdevito is describing here: in NCCL 2.19.4+ we've observed that calling `ncclCommAbort` when other (intra-node) ranks don't also abort will also hang, and the response from NCCL is that while this was allowed for a number of versions, they are saying this is once again the expected behavior. In fewer words, I'm not sure this particular point (`_shutdown` can nominally hang while `ncclCommAbort` should not) will be true with current NCCL versions.

CC @KaimingOuyang to chime in

Reply:
well keep in mind that users may use 2.17 for a while - unless we want to be really aggressive and mark that as unsupported.

so it seems like a desirable feature to have a timeout for ncclCommAbort, but since we can't rely on that being in older nccl verisons, we may have to build it in the pytorch layer for the time being.

Reply:
I agree ncclCommAbort hanging is not pleasant behavior, unfortunately, this hang is limited by CUDA and we had no choice. For now, we have strict semantics that every active rank must call abort. However, it is possible to lift it by using CUMEM API and loosen the abort semantics in the future.

For old NCCL version, it does not hang since we directly free everything during abort, this can cause severe issue like segfault, but somehow, we got lucky and didn't meet the issue.

Reply:
@KaimingOuyang "For old NCCL version, it does not hang since we directly free everything during abort", what is old NCCL version exactly? We can repro the hang consistently with version 2.17.

BTW, I got a working version of calling abort synchronously with a timeout in Pytorch layer, would update the PR soon after I clean up other stuffs.

Reply:
Should be 2.14.x

Reply:
Updated the PR to make abort() async with a timeout, throwing exception when timeout happens. Also add an env to gate the feature rollout.